src/plugins/KeyboardMapping.js

```diff
@@ -0,0 +1,748 @@
+            fn: (config) => {
+                // check it's a valid JSON
+                try {
+                    this.state.mappedKeysConfig = JSON.parse(JSON.stringify(config));
+                } catch (err) {
+                    throw new Error('Invalid JSON');
+                }
+                this.state.config = config;
+            },
+            description: 'Submit a config for mapping keys',
+        });
+
+        // active trace when click on screen
+        this.instance.apiManager.registerFunction({
+            name: 'activeKeyMappingDebug',
+            category: 'keyMapping',
+            fn: (isTraceActivate = false, isGridActivate = false) => {
+                this.activateTrace(isTraceActivate);
+                this.activateGrid(isGridActivate);
+            },
+            description: `Activate debug mode for key mapping. the first parameter activate 
+            feature "click on screen add a div with x, y and x%, y%coordonates.\n
+            The second parameter activate a grid on the screen to help mapping keys. 
+            10% of the screen width and height.`,
+        });
+
+        // active trace when click on screen
+        this.instance.apiManager.registerFunction({
+            name: 'enable',
+            category: 'keyMapping',
+            fn: (isActive = false) => {
+                this.state.isActive = isActive;
+            },
+            description: 'Activate the keyMapping plugin.',
+        });
+    }
+
+    renderToolbarButton() {
+        const toolbars = this.instance.getChildByClass(this.instance.root, 'gm-toolbar');
+        if (!toolbars) {
+            return; // if we don't have toolbar, we can't spawn the widget
+        }
+
+        const toolbar = toolbars.children[0];
+        this.toolbarBtn = document.createElement('li');
+        this.toolbarBtnImage = document.createElement('div');
+        this.toolbarBtnImage.className = 'gm-icon-button gm-keymapping-button';
+        this.toolbarBtnImage.title = this.i18n.KEYMAPPING_TITLE || 'Key Mapping';
+        this.toolbarBtn.onclick = () => (this.state.isActive = !this.state.isActive);
+        this.toolbarBtn.appendChild(this.toolbarBtnImage);
+        toolbar.appendChild(this.toolbarBtn);
+    }
+
+    activatePlugin() {
+        if (this.state.isActive) {
+            this.toolbarBtnImage.classList.add('gm-active');
+            this.addKeyboardCallbacks();
+            this.instance.store.dispatch({type: 'KEYBOARD_EVENTS_ENABLED', payload: false});
+            this.instance.store.dispatch({type: 'MOUSE_EVENTS_ENABLED', payload: false});
+        } else {
+            this.toolbarBtnImage.classList.remove('gm-active');
+            this.removeKeyboardCallbacks();
+            this.instance.store.dispatch({type: 'KEYBOARD_EVENTS_ENABLED', payload: !this.state.isPaused});
+            this.instance.store.dispatch({type: 'MOUSE_EVENTS_ENABLED', payload: !this.state.isPaused});
+        }
+    }
+
+    setupMappedKeysConfig() {
```

Comment:
maybe add a doc comment explaining that this simply flattens the structure, and what `mappedKeysConfig` and `workingMappedKeysConfig` are?